### DISCUSSION TITLE
A List relation is not allowed on module tables.

### DISCUSSION THREAD

**Discussion by KristijanZic** [serverpod/serverpod#2792] [Q&A] [answered]
So I'm trying to decouple my server a little bit and separate some models into modules but then when I try to use a model from a module as a List<module:Model> I get the following error:

```
Found 1 issue.
Error on line 15, column 47 of lib/src/models/warehouse/top_shelf.spy.yaml: A List relation is not allowed on module tables.
  ╷
1 │ relation
  │ ^^^^^^^^
```

relation in the top_shelf:
```
products: List<module:products:ProductDTO?>?, relation(optional)
```

Def:
```
class: ProductDTO
table: products_product

fields:
  # product type
  type: ProductType
```
  
Is there a workaround for this, am I doing something wrong or is there just no support for this in which case I'd like to turn this into a feature request and get some info on the timeline for this feature?

**Comment by SandPod** [comment, marked as answer]
### TLDR
Defining a list relation directly to a module model is not allowed. This would require changing the model in the module, thus breaking the encapsulation of the module models.

Workaround documented here: https://docs.serverpod.dev/concepts/database/relations/modules#advanced-example

### Relations are defined
When a relation is defined, this connection needs to be preserved some way in the database. This is done by storing the primary key of the related record in the other table. So Table A will get an additional column that stores the primary key of the associated record in Table B.

### How list relations are defined 
When a list relation is defined, a single record from Table A should now be associated with multiple records from Table B. The way this is solved in the database is by modifying Table B to have an extra column that stores the primary key of the record associated with it in Table A.

So, if you create a list relation towards a module Model, this would require a modification of that module's database definition, breaking the encapsulation of the module.

### Workaround
To get around this issue and preserve the module encapsulation, you can define a connection model (or junction table) in your project. 

This is documented here: https://docs.serverpod.dev/concepts/database/relations/modules#advanced-example

### Feature status
Breaking encapsulation for modules is not something we are planning to implement.
Will transform this into a Q&A discussion and mark this as an answer.

**Reply by KristijanZic** [reply]
Thank you, this settles it for me 👍
